capsules/extra/src/nonvolatile_storage_driver.rs

```diff
@@ -202,9 +450,485 @@ impl<'a> NonvolatileStorage<'a> {
+        region_header_address: usize,
+    ) -> Result<(), ErrorCode> {
+        if DEBUG {
+            debug!(
+                "[NONVOLATILE_STORAGE_DRIVER]: Writing region header to {:#x}",
+                region_header_address
+            );
+        }
+
+        let owner_slice = u32_to_u8_slice(region_header.shortid);
+        let length_slice = usize_to_u8_slice(region_header.length);
+
+        self.header_buffer.map_or(Err(ErrorCode::NOMEM), |buffer| {
+            // copy owner to static buffer
+            for (i, c) in buffer[0..owner_slice.len()].iter_mut().enumerate() {
+                *c = owner_slice[i];
+            }
+            // copy length to static buffer
+            for (i, c) in buffer[owner_slice.len()..REGION_HEADER_LEN]
+                .iter_mut()
+                .enumerate()
+            {
+                *c = length_slice[i];
+            }
+            Ok(())
+        })?;
+
+        let region = AppRegion {
+            offset: region_header_address + REGION_HEADER_LEN,
+            length: region_header.length,
+        };
+
+        self.enqueue_command(
+            NonvolatileCommand::HeaderWrite(HeaderWriteAction::WritingRegionHeader(
+                processid, region,
+            )),
+            region_header_address,
+            owner_slice.len() + length_slice.len(),
+            None,
+        )
+    }
+
+    fn header_read_done(&self, action: HeaderReadAction) -> Result<(), ErrorCode> {
+        match action {
+            HeaderReadAction::ReadingMagicHeader => {
+                let magic_header = self.header_buffer.map_or(Err(ErrorCode::NOMEM), |buffer| {
+                    // copy first few bytes from static buffer to local slice
+                    let mut magic_header_slice = [0; MAGIC_HEADER_LEN];
+                    for (i, c) in buffer[0..magic_header_slice.len()].iter().enumerate() {
+                        magic_header_slice[i] = *c;
+                    }
+                    Ok(u8_slice_to_u32(&magic_header_slice))
+                })?;
+
+                // check validity of magic header read from storage
+                if magic_header != MAGIC_HEADER {
+                    // if the magic header not correct, assume this capsule has never been
+                    // run and write the correct header to storage
+                    self.write_magic_header()
+                } else {
+                    Ok(())
+                }
+            }
+            HeaderReadAction::ReadingRegionHeader(region_header_address) => {
+                // copy first few bytes from static buffer to local slice
+                let mut header_slice = [0; REGION_HEADER_LEN];
+                self.header_buffer.map_or(Err(ErrorCode::NOMEM), |buffer| {
+                    for (i, c) in buffer[0..header_slice.len()].iter().enumerate() {
+                        header_slice[i] = *c;
+                    }
+                    Ok(())
+                })?;
```

Comment:
This shouldn't be necessary.

Also if it is helpful this file https://github.com/tock/tock/blob/master/libraries/tock-tbf/src/types.rs does a lot of converting buffers to various types.

Reply:
I was able to replace most of the type conversions with bulitin rust functions (ex: `u32::from_le_bytes`).

However, these functions require a slice with a constant length. The buffer I get back from the closure has a type of `&mut [u8]` instead of something like `&mut [u8; 4]`. So right now I'm constructing a constant length buffer of the right size and copying the data over to it manually: 

https://github.com/tock/tock/blob/d02e6325cc22b2eeb101dac1dbdfd6263c2636a2/capsules/extra/src/nonvolatile_storage_driver.rs#L671-L673

Reply:
`&mut [u8]` implements [`TryInto<[u8; 4]>`](https://doc.rust-lang.org/stable/std/convert/trait.TryInto.html), so you should be able to write something like:

```rust
let owner = u32::from_le_bytes(owner_slice.try_into().unwrap());
```

Reply:
Changed here: https://github.com/tock/tock/pull/4109/commits/a3f62a066ecd0a187efa2365584488f5ac3bb8a2. Nice to remove that extra copy!